common/network-shuffle/src/main/java/org/apache/spark/network/shuffle/RemoteBlockPushResolver.java

```diff
@@ -0,0 +1,915 @@
+              writeAnyDeferredBlocks();
+            }
+            long updatedPos = partitionInfo.getPosition() + length;
+            boolean indexUpdated = false;
+            if (updatedPos - partitionInfo.getLastChunkOffset() >= minChunkSize) {
+              partitionInfo.updateChunkInfo(updatedPos, mapId);
+              indexUpdated = true;
+            }
+            partitionInfo.setPosition(updatedPos);
+            partitionInfo.setCurrentMapId(-1);
+
+            // update merged results
+            partitionInfo.blockMerged(mapId);
+            if (indexUpdated) {
+              partitionInfo.resetChunkTracker();
+            }
+          } else {
+            deferredBufs = null;
+            throw new RuntimeException(String.format("%s %s to merged shuffle",
+                ErrorHandler.BlockPushErrorHandler.BLOCK_APPEND_COLLISION_DETECTED_MSG_PREFIX,
+                msg.blockId));
+          }
+        }
+        startBlockWrite = false;
+      }
+
+      @Override
+      public void onFailure(String streamId, Throwable throwable) throws IOException {
+        if (errorHandler.shouldLogError(throwable)) {
+          logger.error("Encountered issue when merging shuffle partition block {}", msg, throwable);
+        } else {
+          logger.debug("Encountered issue when merging shuffle partition block {}", msg, throwable);
+        }
+        // Only update partitionInfo if the failure corresponds to a valid request. If the
+        // request is too late, i.e. received after shuffle merge finalize, #onFailure will
+        // also be triggered, and we can just ignore. Also, if we couldn't find an opportunity
+        // to write the block data to disk, we should also ignore here.
+        if (startBlockWrite && partitionInfo != null && partitions.containsKey(partitionId)) {
+          synchronized (partitionInfo) {
+            logger.debug("{} shuffleId {} reduceId {} set encountered failure", partitionId.appId,
+                partitionId.shuffleId, partitionId.reduceId);
+            partitionInfo.setCurrentMapId(-1);
+            partitionInfo.setEncounteredFailure(true);
+          }
+        }
+      }
+    };
+  }
+
+  @Override
+  public MergeStatuses finalizeShuffleMerge(FinalizeShuffleMerge msg) throws IOException {
+    logger.info("Finalizing shuffle {} from Application {}.", msg.shuffleId, msg.appId);
+    List<RoaringBitmap> bitmaps = new LinkedList<>();
+    List<Integer> reduceIds = new LinkedList<>();
+    List<Long> sizes = new LinkedList<>();
+    Iterator<Map.Entry<AppShufflePartitionId, AppShufflePartitionInfo>> iterator =
+        partitions.entrySet().iterator();
+    while (iterator.hasNext()) {
+      Map.Entry<AppShufflePartitionId, AppShufflePartitionInfo> entry = iterator.next();
+      AppShufflePartitionId partitionId = entry.getKey();
+      AppShufflePartitionInfo partition = entry.getValue();
+      if (partitionId.compareAppShuffleId(msg.appId, msg.shuffleId)) {
```

Comment:
When the shuffle gets finalized, we want to close all the reducer partitions belonging to that shuffle. The `appShufflePartitionInfo` encapsulates all the relevant info of a particular reducer partition.
Here we need all the `appShufflePartitionInfos` that belong to the shuffle that is finalized. This  may have been easier to read if there was a mapping from `(appId, shuffleId) -> list(appShufflePartitionInfo)`.  This adds to the memory of this shuffle service.
Would adding a comment here help?

Reply:
I understand what it does here. But I'm thinking it might be simpler if we could have that map. At least, we don't need to traverse the whole `partitions` to find all the reducer partitions belonging to a certain partition. We can just get them by `map.get(shuffleId)`. The time complexity here is O(n) vs O(1).

Reply:
Since shuffle id will conflict across applications, I want to understand the proposal better @Ngone51. Is it to replace `partitions` with`Map[(application_id, shuffleId) -> Map [reduceId -> AppShufflePartitionInfo] ]` ?

Reply:
Oh yeah. Sorry, I forget to mention the `applicationId`. And yes the `Map[(application_id, shuffleId) -> Map [reduceId -> AppShufflePartitionInfo]]` is something that I'm proposing.

Reply:
I am working on replacing `partitions` with `Map[(application_id, shuffleId) -> Map [reduceId -> AppShufflePartitionInfo]`

Reply:
I have made this change so resolving this.